### PR TITLE
Transformed priors

### PR DESCRIPTION
**Pull request by barkls**
Documentation on using these new things is lacking. We should have a user guides on Model & Prior objects, and also a tutorial using the new features. I'm happy to get this merged in first though

### CODE REVIEW

**Comment by barkls** [comment]
oh no! I forgot to update the changelog. Don't merge until I do...

File: holopy/inference/model.py
Comment (briandleahy) [inline holopy/inference/model.py:69]: What happened to `make_complex`? Is the idea that now we say `n=prior.Gaussian(1.59, 0.1) + 1j * prior.Uniform(0, 1e-3)`?

Comment (briandleahy) [reply holopy/inference/model.py:69]: (I'm mostly asking about documentation.)

File: holopy/inference/prior.py
Comment (briandleahy) [inline holopy/inference/prior.py:90]: Just curious -- why are these not `return TransformedPrior(operator.truediv, [self, value])` and `return TransformedPrior(operator.truediv, [value, self])`?

File: holopy/inference/prior.py
Comment (briandleahy) [inline holopy/inference/prior.py:351]: Nice!! :+1:

File: holopy/inference/prior.py
Comment (briandleahy) [inline holopy/inference/prior.py:41]: It might be nice to add a list of methods that `Prior` subclasses are expected implement, for the future devs who may extend these. I think either adding a list in the `Prior` docstring or adding empty methods, like
```python
    def lnprob(self, p)
        raise NotImplementedError("implement in subclass")
```
works. This might be out of the scope of this PR though.

**Review by briandleahy — APPROVED** [review]
Looks good to me @barkls . I left a few minor comments, but I don't think they are necessary for this PR. 

At some point we do need to update the docs for this, but that should probably be a separate PR.

Let me know when you want me to merge this in (since you said you wanted to update the change log).

File: holopy/inference/model.py
Comment (barkls) [reply holopy/inference/model.py:69]: `make_complex` was just a dispatch function used by a Model to unpack a map (e.g. into a scatterer). It became redundant when `ComplexPrior` was rolled into `TransformedPrior` which has its own similar function.

File: holopy/inference/prior.py
Comment (barkls) [reply holopy/inference/prior.py:90]: They probably could be - previously the math behaviour had to be written manually so I tried to use identities to get everything onto `__mul__` and `__add__`. That's less of an issue now, but there is still specific handling for special values like 0 or 1. Do you think it would be more clear to use `operator` functions throughout?

File: holopy/inference/prior.py
Comment (barkls) [reply holopy/inference/prior.py:351]: Nope, the most important `TransformedPrior` methods (`guess` and `sample`) are inherited directly.

I do override some placeholders (`prob`, `lnprob`, `map_keys`) in `TransformedPrior` with more intelligent behaviour specific to the `ComplexPrior`. I do think there's a future in which other types of transformations get their own classes (e.g. arctan or other trig), and they would need to override these defaults as well. Let me know if there's a better way to do this.

Finally, there are some properties (`real` and `imag`) that are just convenience ways of accessing `base_prior` components that also maintain backwards compatibility. Actually `map_keys` was added to maintain backwards compatibility as well. I initially had `Model` just follow the `TransformedPrior` behaviour without asking the object, but that didn't match previous parameter names for `ComplexPrior` objects.

File: holopy/inference/prior.py
Comment (barkls) [reply holopy/inference/prior.py:41]: I'm planning to write a user guide for the `Prior` class at some point, like the ones for `Scatterer` and `ScatteringTheory`. Would that be sufficient or do you think I should put something in the code as well?

File: holopy/inference/prior.py
Comment (briandleahy) [reply holopy/inference/prior.py:90]: Personally, I think that keeping the same pattern throughout would be the clearest, but I'm not sure. For instance, if you are end up doing a lot of error handling, it could be cleaner to call the `__mul__` like you are doing here. 

It's up to you.

File: holopy/inference/prior.py
Comment (briandleahy) [reply holopy/inference/prior.py:351]: :+1:

File: holopy/inference/prior.py
Comment (briandleahy) [reply holopy/inference/prior.py:41]: I think something in the code as well would be good. (Just because someone could update the code in the future and forget to update the user guide.) Again, not super urgent though.

**Comment by barkls** [comment]
Thanks for the review @briandleahy! Ready to merge (assuming tests pass)

**Pull request merged**
